In [1]:
%%capture
!pip install psycopg2-binary

%load_ext sql

%sql postgresql+psycopg2://postgres:7089@localhost:5432/banknifty

This project presents a comprehensive analysis of the Bank Nifty index, focusing on data from the National Stock Exchange (NSE) spanning five years, from 2019 to 2023. The Bank Nifty index, which comprises the most liquid and large-cap banking stocks, serves as a crucial barometer for the Indian banking sector's performance.

Our analysis delves into various aspects of the index, including key metrics such as share traded, turnover, and price fluctuations. By examining these metrics, we aim to uncover trends and patterns that offer insights into the behavior and performance of the index over the past five years.

Key analyses include:

Correlation Analysis: Assessing relationships between different financial metrics, such as share traded versus turnover and price volatility.

Trend Analysis: Evaluating average values of critical parameters like open, low, high, and close amounts on a monthly basis.

Velocity Measures: Calculating the velocity of understand the difference between highest and lowest price of the day for following 5 consicutive years.


In [13]:
%sql SELECT * FROM banknifty5yr limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
10 rows affected.


[(datetime.date(2019, 1, 1), Decimal('27231.4'), Decimal('27430.55'), Decimal('27019.05'), Decimal('27392.4'), Decimal('116379065'), Decimal('3497.39')),
 (datetime.date(2019, 1, 2), Decimal('27297'), Decimal('27397.55'), Decimal('27077.6'), Decimal('27174.7'), Decimal('176599164'), Decimal('4422.44')),
 (datetime.date(2019, 1, 3), Decimal('27181.6'), Decimal('27206.2'), Decimal('26923.75'), Decimal('26959.85'), Decimal('151201387'), Decimal('4032.83')),
 (datetime.date(2019, 1, 4), Decimal('26999.7'), Decimal('27274.5'), Decimal('26926.1'), Decimal('27195'), Decimal('189451222'), Decimal('4440.79')),
 (datetime.date(2019, 1, 7), Decimal('27378.65'), Decimal('27477.8'), Decimal('27279.55'), Decimal('27304.55'), Decimal('161026470'), Decimal('4112.84')),
 (datetime.date(2019, 1, 8), Decimal('27301.9'), Decimal('27542.4'), Decimal('27161.95'), Decimal('27509.5'), Decimal('150550286'), Decimal('4359.43')),
 (datetime.date(2019, 1, 9), Decimal('27651.25'), Decimal('27754.45'), Decimal('27409.5'), Decimal('27720.4'), Decimal('199818392'), Decimal('6148.74')),
 (datetime.date(2019, 1, 10), Decimal('27713.55'), Decimal('27713.55'), Decimal('27488.2'), Decimal('27528.55'), Decimal('139266179'), Decimal('4308.94')),
 (datetime.date(2019, 1, 11), Decimal('27602.8'), Decimal('27612.45'), Decimal('27382.45'), Decimal('27453.9'), Decimal('116552104'), Decimal('3510.63')),
 (datetime.date(2019, 1, 14), Decimal('27389.2'), Decimal('27389.8'), Decimal('27182.2'), Decimal('27248.25'), Decimal('212863498'), Decimal('4859.32'))]

##### Velocity Measure

In [14]:
%sql SELECT date, high_amt - low_amt AS velocity FROM banknifty5yr limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
10 rows affected.


[(datetime.date(2019, 1, 1), Decimal('411.50')),
 (datetime.date(2019, 1, 2), Decimal('319.95')),
 (datetime.date(2019, 1, 3), Decimal('282.45')),
 (datetime.date(2019, 1, 4), Decimal('348.4')),
 (datetime.date(2019, 1, 7), Decimal('198.25')),
 (datetime.date(2019, 1, 8), Decimal('380.45')),
 (datetime.date(2019, 1, 9), Decimal('344.95')),
 (datetime.date(2019, 1, 10), Decimal('225.35')),
 (datetime.date(2019, 1, 11), Decimal('230.00')),
 (datetime.date(2019, 1, 14), Decimal('207.6'))]

![Velocity Trend](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/c9b5a9dcdd99923943db4b96bd3c209c4ce810b1/Velocity%20Trend.png)

In [15]:
%sql select * ,extract (year from date) as year,extract (month from date)	as month from banknifty5yr limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
10 rows affected.


[(datetime.date(2019, 1, 1), Decimal('27231.4'), Decimal('27430.55'), Decimal('27019.05'), Decimal('27392.4'), Decimal('116379065'), Decimal('3497.39'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 2), Decimal('27297'), Decimal('27397.55'), Decimal('27077.6'), Decimal('27174.7'), Decimal('176599164'), Decimal('4422.44'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 3), Decimal('27181.6'), Decimal('27206.2'), Decimal('26923.75'), Decimal('26959.85'), Decimal('151201387'), Decimal('4032.83'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 4), Decimal('26999.7'), Decimal('27274.5'), Decimal('26926.1'), Decimal('27195'), Decimal('189451222'), Decimal('4440.79'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 7), Decimal('27378.65'), Decimal('27477.8'), Decimal('27279.55'), Decimal('27304.55'), Decimal('161026470'), Decimal('4112.84'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 8), Decimal('27301.9'), Decimal('27542.4'), Decimal('27161.95'), Decimal('27509.5'), Decimal('150550286'), Decimal('4359.43'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 9), Decimal('27651.25'), Decimal('27754.45'), Decimal('27409.5'), Decimal('27720.4'), Decimal('199818392'), Decimal('6148.74'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 10), Decimal('27713.55'), Decimal('27713.55'), Decimal('27488.2'), Decimal('27528.55'), Decimal('139266179'), Decimal('4308.94'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 11), Decimal('27602.8'), Decimal('27612.45'), Decimal('27382.45'), Decimal('27453.9'), Decimal('116552104'), Decimal('3510.63'), Decimal('2019'), Decimal('1')),
 (datetime.date(2019, 1, 14), Decimal('27389.2'), Decimal('27389.8'), Decimal('27182.2'), Decimal('27248.25'), Decimal('212863498'), Decimal('4859.32'), Decimal('2019'), Decimal('1'))]

##### Yearly-Highs

In [12]:
%sql SELECT  EXTRACT(year FROM date) AS year,  MAX(high_amt) as max_high_amt from banknifty5yr group by year order by year;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
5 rows affected.


[(Decimal('2019'), Decimal('32613.1')),
 (Decimal('2020'), Decimal('32465.45')),
 (Decimal('2021'), Decimal('41829.6')),
 (Decimal('2022'), Decimal('44151.8')),
 (Decimal('2023'), Decimal('48636.45'))]

##### Yearly Trend Analysis

In [5]:
%sql SELECT extract(year from date) AS year,  ROUND(avg(share_traded), 2) AS sharetraded_avg,  ROUND(avg(turnover_cr), 2) AS turnoverCr_avg, ROUND(avg(open_amt), 2) AS openamt_avg,  ROUND(avg(low_amt), 2) AS lowamt_avg, ROUND(avg(close_amt), 2) AS closeamt_avg, ROUND(avg(high_amt), 2) AS highamt_avg FROM banknifty5yr GROUP BY year ORDER BY year ASC;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
5 rows affected.


[(Decimal('2019'), Decimal('307005222.98'), Decimal('6443.72'), Decimal('29385.37'), Decimal('29144.78'), Decimal('29360.10'), Decimal('29581.08')),
 (Decimal('2020'), Decimal('384173893.47'), Decimal('9070.08'), Decimal('24606.34'), Decimal('24223.86'), Decimal('24558.08'), Decimal('24894.47')),
 (Decimal('2021'), Decimal('295499542.99'), Decimal('7497.16'), Decimal('35418.10'), Decimal('35061.90'), Decimal('35380.77'), Decimal('35699.81')),
 (Decimal('2022'), Decimal('199530498.87'), Decimal('5937.45'), Decimal('37955.79'), Decimal('37661.27'), Decimal('37979.21'), Decimal('38260.29')),
 (Decimal('2023'), Decimal('203428579.08'), Decimal('8769.18'), Decimal('43607.26'), Decimal('43330.82'), Decimal('43573.06'), Decimal('43807.41'))]

![Yearly Trend](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/e63d821a61ba176104a5c085b87e2fe9c09a78fb/Yearly%20Trend.png)

##### Monthly Trend Analysis

In [6]:
%sql SELECT extract(month from date) AS month, ROUND(avg(share_traded), 2) AS sharetraded_avg, ROUND(avg(turnover_cr), 2) AS turnoverCr_avg, ROUND(avg(open_amt), 2) AS openamt_avg, ROUND(avg(low_amt), 2) AS lowamt_avg, ROUND(avg(close_amt), 2) AS closeamt_avg, ROUND(avg(high_amt), 2) AS highamt_avg FROM banknifty5yr GROUP BY month ORDER BY month ASC;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
12 rows affected.


[(Decimal('1'), Decimal('266410449.40'), Decimal('6681.92'), Decimal('33956.72'), Decimal('33644.88'), Decimal('33902.18'), Decimal('34168.92')),
 (Decimal('2'), Decimal('295535114.99'), Decimal('7340.84'), Decimal('34457.89'), Decimal('34130.59'), Decimal('34444.20'), Decimal('34760.62')),
 (Decimal('3'), Decimal('311872974.81'), Decimal('8227.95'), Decimal('32502.87'), Decimal('32027.89'), Decimal('32433.97'), Decimal('32843.34')),
 (Decimal('4'), Decimal('249134388.11'), Decimal('8466.13'), Decimal('32145.70'), Decimal('31795.64'), Decimal('32134.83'), Decimal('32438.69')),
 (Decimal('5'), Decimal('284169853.78'), Decimal('8097.59'), Decimal('32490.11'), Decimal('32183.31'), Decimal('32465.88'), Decimal('32770.63')),
 (Decimal('6'), Decimal('270333201.53'), Decimal('7146.06'), Decimal('32949.12'), Decimal('32668.41'), Decimal('32914.22'), Decimal('33181.47')),
 (Decimal('7'), Decimal('247263745.49'), Decimal('7245.62'), Decimal('33434.53'), Decimal('33174.16'), Decimal('33396.19'), Decimal('33633.18')),
 (Decimal('8'), Decimal('266698834.13'), Decimal('7434.00'), Decimal('33920.37'), Decimal('33650.00'), Decimal('33920.01'), Decimal('34150.24')),
 (Decimal('9'), Decimal('260853915.16'), Decimal('7289.15'), Decimal('34468.44'), Decimal('34166.48'), Decimal('34458.42'), Decimal('34745.42')),
 (Decimal('10'), Decimal('309071168.56'), Decimal('7347.46'), Decimal('34935.62'), Decimal('34634.78'), Decimal('34910.01'), Decimal('35192.33')),
 (Decimal('11'), Decimal('286752434.47'), Decimal('8182.55'), Decimal('36798.70'), Decimal('36538.91'), Decimal('36810.10'), Decimal('37025.16')),
 (Decimal('12'), Decimal('292168411.23'), Decimal('7296.11'), Decimal('37634.82'), Decimal('37344.63'), Decimal('37615.94'), Decimal('37849.93'))]

![Monthly Trend](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/e63d821a61ba176104a5c085b87e2fe9c09a78fb/Monthly%20Trend.png)

##### Correlation Analysis

In [7]:
%sql select corr (share_traded, turnover_cr) as traded_vs_turnover, corr (low_amt, share_traded) as low_amt_vs_traded, corr (high_amt, share_traded) as high_amt_vs_traded, corr (low_amt, turnover_cr) as low_amt_vs_turnover, corr (high_amt, turnover_cr) as high_amt_vs_turnover, corr (open_amt, close_amt) as openamt_vs_closeamt, corr (open_amt, high_amt) as openamt_vs_highamt from banknifty5yr;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
1 rows affected.


[(0.6584548803621242, -0.3984282277194848, -0.38417398824249077, -0.06391085144557071, -0.04556129277423231, 0.9985604338283833, 0.9993236051773501)]

The analysis reveals that the relationship between volume and turnover is moderately positive, indicating that these metrics tend to increase together. However, the correlation between prices (low/high) and traded volume or turnover is weak, suggesting that other factors might be influencing these variables.

The exceptionally high correlations between open-close and open-high prices suggest that these prices are highly predictable based on the opening price of the day, implying a certain level of market stability or low intra-day volatility for the Bank Nifty on the analyzed days.

Overall, this analysis provides insights into the interactions between different financial metrics, offering a basis for further exploration or decision-making based on these correlations.


###### CORRELATION RANGE
r=1: Perfect positive correlation.

0.7<r<1: Strong positive correlation.

0<r≤0.7: Weak to moderate positive correlation.

r=0: No correlation.

−0.7<r<0: Weak to moderate negative correlation.

r=−1: Perfect negative correlation.

##### True Range

In [16]:
%sql SELECT date,close_amt,high_amt, low_amt, GREATEST(high_amt - low_amt, high_amt - LAG(close_amt) OVER (ORDER BY date),low_amt - LAG(open_amt) OVER (ORDER BY date)) AS true_range FROM banknifty5yr ORDER BY true_range DESC limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
10 rows affected.


[(datetime.date(2020, 3, 13), Decimal('25166.45'), Decimal('25670.75'), Decimal('21351.7'), Decimal('4319.05')),
 (datetime.date(2019, 9, 23), Decimal('30566.2'), Decimal('30801.3'), Decimal('29776.3'), Decimal('2897.65')),
 (datetime.date(2021, 2, 1), Decimal('33089.05'), Decimal('33305.3'), Decimal('30906.45'), Decimal('2739.8')),
 (datetime.date(2019, 9, 20), Decimal('28981.55'), Decimal('29418.95'), Decimal('26727.55'), Decimal('2691.40')),
 (datetime.date(2021, 2, 2), Decimal('34267.9'), Decimal('34652.5'), Decimal('33583.15'), Decimal('2606.80')),
 (datetime.date(2023, 2, 1), Decimal('40513'), Decimal('42015.65'), Decimal('39490.5'), Decimal('2525.15')),
 (datetime.date(2020, 3, 19), Decimal('20083.5'), Decimal('21035.05'), Decimal('18675.65'), Decimal('2359.40')),
 (datetime.date(2020, 3, 18), Decimal('20580.2'), Decimal('22544.05'), Decimal('20267.95'), Decimal('2276.10')),
 (datetime.date(2020, 3, 25), Decimal('18481.05'), Decimal('18740.2'), Decimal('16635.5'), Decimal('2104.7')),
 (datetime.date(2020, 3, 23), Decimal('16917.65'), Decimal('18895.65'), Decimal('16791.05'), Decimal('2104.60'))]

![True Range](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/62413306a5ba8ac5e7e1f6cf45388d8ec1bc8924/True%20Range.png)

The True Range of Bank Nifty over the past five years, indicating periods of high volatility, particularly around early 2020. The peak value of 4,319 suggests a dramatic market response, likely influenced by external factors such as economic events or market crises. Overall, the trend shows a gradual stabilization in True Range values post-2020, reflecting a return to more typical market conditions.


###### True Range:
True Range is a measure of market volatility used in technical analysis. It's calculated as the greatest of the following:

1. The difference between the current day's high and low

2. The absolute difference between the current day's high and the previous day's close

3. The absolute difference between the current day's low and the previous day's close

##### Simple Moving Average (SMA)

In [17]:
%sql select date, close_amt, avg (close_amt) over (order by date rows between 9 preceding and current row) as moving_average_close, open_amt, avg (open_amt) over (order by date rows between 9 preceding and current row) as moving_average_open, low_amt, avg (low_amt) over (order by date rows between 9 preceding and current row) as moving_average_low, high_amt, avg (high_amt) over (order by date rows between 9 preceding and current row) as moving_average_high from banknifty5yr order by date limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/banknifty
10 rows affected.


[(datetime.date(2019, 1, 1), Decimal('27392.4'), Decimal('27392.400000000000'), Decimal('27231.4'), Decimal('27231.400000000000'), Decimal('27019.05'), Decimal('27019.050000000000'), Decimal('27430.55'), Decimal('27430.550000000000')),
 (datetime.date(2019, 1, 2), Decimal('27174.7'), Decimal('27283.550000000000'), Decimal('27297'), Decimal('27264.200000000000'), Decimal('27077.6'), Decimal('27048.325000000000'), Decimal('27397.55'), Decimal('27414.050000000000')),
 (datetime.date(2019, 1, 3), Decimal('26959.85'), Decimal('27175.650000000000'), Decimal('27181.6'), Decimal('27236.666666666667'), Decimal('26923.75'), Decimal('27006.800000000000'), Decimal('27206.2'), Decimal('27344.766666666667')),
 (datetime.date(2019, 1, 4), Decimal('27195'), Decimal('27180.487500000000'), Decimal('26999.7'), Decimal('27177.425000000000'), Decimal('26926.1'), Decimal('26986.625000000000'), Decimal('27274.5'), Decimal('27327.200000000000')),
 (datetime.date(2019, 1, 7), Decimal('27304.55'), Decimal('27205.300000000000'), Decimal('27378.65'), Decimal('27217.670000000000'), Decimal('27279.55'), Decimal('27045.210000000000'), Decimal('27477.8'), Decimal('27357.320000000000')),
 (datetime.date(2019, 1, 8), Decimal('27509.5'), Decimal('27256.000000000000'), Decimal('27301.9'), Decimal('27231.708333333333'), Decimal('27161.95'), Decimal('27064.666666666667'), Decimal('27542.4'), Decimal('27388.166666666667')),
 (datetime.date(2019, 1, 9), Decimal('27720.4'), Decimal('27322.342857142857'), Decimal('27651.25'), Decimal('27291.642857142857'), Decimal('27409.5'), Decimal('27113.928571428571'), Decimal('27754.45'), Decimal('27440.492857142857')),
 (datetime.date(2019, 1, 10), Decimal('27528.55'), Decimal('27348.118750000000'), Decimal('27713.55'), Decimal('27344.381250000000'), Decimal('27488.2'), Decimal('27160.712500000000'), Decimal('27713.55'), Decimal('27474.625000000000')),
 (datetime.date(2019, 1, 11), Decimal('27453.9'), Decimal('27359.872222222222'), Decimal('27602.8'), Decimal('27373.094444444444'), Decimal('27382.45'), Decimal('27185.350000000000'), Decimal('27612.45'), Decimal('27489.938888888889')),
 (datetime.date(2019, 1, 14), Decimal('27248.25'), Decimal('27348.710000000000'), Decimal('27389.2'), Decimal('27374.705000000000'), Decimal('27182.2'), Decimal('27185.035000000000'), Decimal('27389.8'), Decimal('27479.925000000000'))]

![Moving Average](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/62413306a5ba8ac5e7e1f6cf45388d8ec1bc8924/Moving%20Average.png)
![SMA close_amt vs Actual Close_amt](https://raw.githubusercontent.com/BhaktiDangra/Bank_nifty_5yr/62413306a5ba8ac5e7e1f6cf45388d8ec1bc8924/SMA%20close_amt%20vs%20Actual%20Close_amt.png)

The relationship between the Moving Average Close and the actual Close Amount of Bank Nifty over time. Both lines show a general upward trend, indicating positive market performance. 

###### Conclusion

The analysis of the Bank Nifty data reveals a steady upward trend in prices over the past five years, indicating strong market performance and growing investor confidence. Despite some periods of volatility, particularly around early 2020, the overall trajectory shows resilience and stability. The velocity and turnover trends further support this growth, highlighting active trading and market dynamics that contribute to the positive outlook.